In [81]:
import re
import pandas as pd

In [82]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\GITESH\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [83]:
messages=pd.read_csv('smsspamcollection/SMSSpamCollection',sep='\t',names=['label','message'])

In [84]:
messages

label                                            message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [85]:
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()
from nltk.corpus import stopwords

This is for BOW

In [86]:
print(len(messages))
corpus=[]
for i in range(0,len(messages)):
    review=re.sub('[^a-zA-z]',' ',messages['message'][i])
    review=review.lower()
    review=review.split()
    message=[ps.stem(word) for word in review if not word in stopwords.words('english')]
    review=' '.join(message)
    corpus.append(review)
corpus

5572


['go jurong point crazi avail bugi n great world la e buffet cine got amor wat',
 'ok lar joke wif u oni',
 'free entri wkli comp win fa cup final tkt st may text fa receiv entri question std txt rate c appli',
 'u dun say earli hor u c alreadi say',
 'nah think goe usf live around though',
 'freemsg hey darl week word back like fun still tb ok xxx std chg send rcv',
 'even brother like speak treat like aid patent',
 'per request mell mell oru minnaminungint nurungu vettam set callertun caller press copi friend callertun',
 'winner valu network custom select receivea prize reward claim call claim code kl valid hour',
 'mobil month u r entitl updat latest colour mobil camera free call mobil updat co free',
 'gonna home soon want talk stuff anymor tonight k cri enough today',
 'six chanc win cash pound txt csh send cost p day day tsandc appli repli hl info',
 'urgent week free membership prize jackpot txt word claim c www dbuk net lccltd pobox ldnw rw',
 'search right word thank breather

This is for Word2Vec alternative steps for preprocessing

In [87]:
from nltk.tokenize import sent_tokenize
from gensim.utils import simple_preprocess

In [88]:
from nltk.stem import WordNetLemmatizer
lm=WordNetLemmatizer()

In [89]:
corpus=[]
for i in range(0,len(messages)):
    review=re.sub('[^a-zA-z]',' ',messages['message'][i])
    review=review.lower()
    review=review.split()
    message=[lm.lemmatize(word) for word in review]
    review=' '.join(message)
    corpus.append(review)
corpus

['go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat',
 'ok lar joking wif u oni',
 'free entry in a wkly comp to win fa cup final tkts st may text fa to to receive entry question std txt rate t c s apply over s',
 'u dun say so early hor u c already then say',
 'nah i don t think he go to usf he life around here though',
 'freemsg hey there darling it s been week s now and no word back i d like some fun you up for it still tb ok xxx std chgs to send to rcv',
 'even my brother is not like to speak with me they treat me like aid patent',
 'a per your request melle melle oru minnaminunginte nurungu vettam ha been set a your callertune for all caller press to copy your friend callertune',
 'winner a a valued network customer you have been selected to receivea prize reward to claim call claim code kl valid hour only',
 'had your mobile month or more u r entitled to update to the latest colour mobile with camera for free call the mobile up

In [90]:
words=[]
for sent in corpus:
    sent_tokens=sent_tokenize(sent)
    for word in sent_tokens:
        words.append(simple_preprocess(word))

In [91]:
words

[['go',
  'until',
  'jurong',
  'point',
  'crazy',
  'available',
  'only',
  'in',
  'bugis',
  'great',
  'world',
  'la',
  'buffet',
  'cine',
  'there',
  'got',
  'amore',
  'wat'],
 ['ok', 'lar', 'joking', 'wif', 'oni'],
 ['free',
  'entry',
  'in',
  'wkly',
  'comp',
  'to',
  'win',
  'fa',
  'cup',
  'final',
  'tkts',
  'st',
  'may',
  'text',
  'fa',
  'to',
  'to',
  'receive',
  'entry',
  'question',
  'std',
  'txt',
  'rate',
  'apply',
  'over'],
 ['dun', 'say', 'so', 'early', 'hor', 'already', 'then', 'say'],
 ['nah',
  'don',
  'think',
  'he',
  'go',
  'to',
  'usf',
  'he',
  'life',
  'around',
  'here',
  'though'],
 ['freemsg',
  'hey',
  'there',
  'darling',
  'it',
  'been',
  'week',
  'now',
  'and',
  'no',
  'word',
  'back',
  'like',
  'some',
  'fun',
  'you',
  'up',
  'for',
  'it',
  'still',
  'tb',
  'ok',
  'xxx',
  'std',
  'chgs',
  'to',
  'send',
  'to',
  'rcv'],
 ['even',
  'my',
  'brother',
  'is',
  'not',
  'like',
  'to',
  'spea

In [92]:
import gensim

Train word2vec from scratch

In [93]:
model=gensim.models.Word2Vec(words)

In [94]:
#To get all the vocalbulary
model.wv.index_to_key

['to',
 'you',
 'the',
 'it',
 'and',
 'in',
 'is',
 'me',
 'my',
 'for',
 'your',
 'call',
 'of',
 'that',
 'have',
 'on',
 'now',
 'are',
 'can',
 'so',
 'but',
 'not',
 'or',
 'we',
 'do',
 'get',
 'at',
 'ur',
 'will',
 'if',
 'be',
 'with',
 'no',
 'just',
 'this',
 'gt',
 'lt',
 'go',
 'how',
 'up',
 'when',
 'ok',
 'day',
 'what',
 'free',
 'from',
 'all',
 'out',
 'know',
 'll',
 'come',
 'like',
 'good',
 'time',
 'am',
 'then',
 'got',
 'wa',
 'there',
 'he',
 'love',
 'text',
 'only',
 'want',
 'send',
 'one',
 'need',
 'txt',
 'today',
 'by',
 'going',
 'don',
 'stop',
 'home',
 'she',
 'about',
 'lor',
 'sorry',
 'see',
 'still',
 'mobile',
 'take',
 'back',
 'da',
 'reply',
 'dont',
 'our',
 'think',
 'tell',
 'week',
 'hi',
 'phone',
 'they',
 'new',
 'please',
 'later',
 'pls',
 'any',
 'her',
 'ha',
 'co',
 'did',
 'been',
 'msg',
 'min',
 'some',
 'an',
 'night',
 'make',
 'dear',
 'who',
 'here',
 'message',
 'say',
 'well',
 'where',
 're',
 'thing',
 'much',
 'oh',

In [95]:
model.wv.similar_by_word('king')

[('cant', 0.9958596229553223),
 ('life', 0.9958307147026062),
 ('hello', 0.9957497715950012),
 ('sweet', 0.9956592321395874),
 ('lot', 0.9955968856811523),
 ('anything', 0.9955640435218811),
 ('smile', 0.9955183267593384),
 ('hear', 0.9954984784126282),
 ('worry', 0.9954607486724854),
 ('may', 0.995426595211029)]

In [96]:
model.wv['king'].shape

(100,)

In [97]:
words[0]

['go',
 'until',
 'jurong',
 'point',
 'crazy',
 'available',
 'only',
 'in',
 'bugis',
 'great',
 'world',
 'la',
 'buffet',
 'cine',
 'there',
 'got',
 'amore',
 'wat']

Convert word2vec to avgword2vev

In [106]:
import numpy as np
def avgword2vec(doc):
    return np.mean([model.wv[word] for word in doc if word in model.wv.index_to_key],axis=0)

In [99]:
!pip install tqdm

In [107]:
from tqdm import tqdm
X=[]
for i in tqdm(range(0,len(words))):
    
    X.append(avgword2vec(words[i]))
X

  0%|          | 0/5569 [00:00<?, ?it/s]d:\Machine Learning\venv\lib\site-packages\numpy\core\fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
d:\Machine Learning\venv\lib\site-packages\numpy\core\_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
100%|██████████| 5569/5569 [00:00<00:00, 10115.82it/s]


[array([-0.1863533 ,  0.27825877,  0.15487093,  0.15889587,  0.08814286,
        -0.49032018,  0.21512018,  0.45334035, -0.30136362, -0.11401383,
        -0.1636093 , -0.37227747, -0.05167628,  0.11206197,  0.19809334,
        -0.16922283,  0.129747  , -0.2776866 , -0.1277856 , -0.52400106,
         0.20513406,  0.10203536,  0.095382  , -0.19607496, -0.05034198,
        -0.01481887, -0.20463373, -0.21207729, -0.2979836 ,  0.04656379,
         0.26518494, -0.00317835,  0.09025203, -0.16799995, -0.12032747,
         0.35798454,  0.0534969 , -0.12466186, -0.13886757, -0.45312688,
         0.05957507, -0.20609148, -0.18205218,  0.02654367,  0.15815495,
        -0.02873812, -0.08505163, -0.02254773,  0.23208365,  0.14045818,
         0.15779005, -0.16119295, -0.04046871,  0.03956752, -0.07469162,
         0.06638633,  0.14005086,  0.01237159, -0.34760895,  0.19131595,
        -0.00225432,  0.1378797 ,  0.00433062, -0.09305862, -0.2590088 ,
         0.27500373,  0.0810816 ,  0.24997726, -0.3

In [108]:
len(X)
#Independent feature

5569

In [ ]:
len(X[0])


100

In [ ]:
## Dependent feature
y=pd.get_dummies(messages['label'])
y

ham   spam
0      True  False
1      True  False
2     False   True
3      True  False
4      True  False
...     ...    ...
5567  False   True
5568   True  False
5569   True  False
5570   True  False
5571   True  False

[5572 rows x 2 columns]

In [ ]:
y=y.iloc[:,0].values
y.shape

(5572,)

As X has 5562 and y Has 5572, 3 are missing so to find that

In [ ]:
print(list(map(lambda x: len(x)>0,corpus)))
y=messages[list(map(lambda x: len(x)>0,corpus))]
print(y)
y=pd.get_dummies(y['label'])
y=y.iloc[:,0].values
y.shape

[True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, Tru

(5569,)

No x and y shape are same and how did we fuond the three are missing like this 

In [ ]:
for i,j,k in zip(list(map(len,corpus)),corpus,messages['message'])

In [ ]:
#Now as the X array is of 100 dimension we need to cioinvert it to the dataframe

In [111]:
import pandas as pd


# Assuming X is a NumPy array or list of NumPy arrays
df = pd.DataFrame([x.reshape(1, -1)[0] for x in X])  # Reshape and flatten each element
df.head


<bound method NDFrame.head of             0         1         2         3         4         5         6   \
0    -0.186353  0.278259  0.154871  0.158896  0.088143 -0.490320  0.215120   
1    -0.174015  0.249120  0.136825  0.141710  0.084993 -0.436311  0.182316   
2    -0.192508  0.301810  0.165681  0.182228  0.064813 -0.524373  0.215520   
3    -0.262483  0.382827  0.208046  0.220823  0.124526 -0.673784  0.293652   
4    -0.226910  0.313297  0.183701  0.182957  0.116667 -0.564499  0.241205   
...        ...       ...       ...       ...       ...       ...       ...   
5564 -0.218926  0.351283  0.203560  0.207763  0.089104 -0.599532  0.263299   
5565 -0.234878  0.335626  0.188674  0.203473  0.112647 -0.599992  0.256242   
5566 -0.256968  0.376677  0.202452  0.206171  0.110097 -0.655870  0.291494   
5567 -0.229128  0.338998  0.187213  0.196091  0.100417 -0.596977  0.257473   
5568 -0.264380  0.361781  0.203385  0.215591  0.127578 -0.651267  0.271769   

            7         8         9

In [ ]:
df.shape

(5569, 100)

In [122]:
df['Output']=y

In [124]:
df.head()

0         1         2         3         4         5         6  \
0 -0.186353  0.278259  0.154871  0.158896  0.088143 -0.490320  0.215120   
1 -0.174015  0.249120  0.136825  0.141710  0.084993 -0.436311  0.182316   
2 -0.192508  0.301810  0.165681  0.182228  0.064813 -0.524373  0.215520   
3 -0.262483  0.382827  0.208046  0.220823  0.124526 -0.673784  0.293652   
4 -0.226910  0.313297  0.183701  0.182957  0.116667 -0.564499  0.241205   

          7         8         9  ...        91        92        93        94  \
0  0.453340 -0.301364 -0.114014  ...  0.170033  0.030420  0.059450  0.438172   
1  0.408362 -0.269871 -0.095602  ...  0.145468  0.024408  0.047345  0.381616   
2  0.444704 -0.314592 -0.147886  ...  0.165275  0.007124  0.045315  0.445852   
3  0.631379 -0.419519 -0.148511  ...  0.233553  0.046094  0.086593  0.595179   
4  0.533170 -0.354483 -0.133306  ...  0.190242  0.038013  0.071998  0.498062   

         95        96        97        98        99  Output  
0  0.159497  0.138176 -0.208883  0.154775  0.022593    True  
1  0.137244  0.123301 -0.193349  0.142204  0.012349    True  
2  0.148977  0.080534 -0.243568  0.179096  0.049865   False  
3  0.222944  0.207379 -0.290350  0.213048  0.020701    True  
4  0.187145  0.166109 -0.257356  0.173881  0.019994    True  

[5 rows x 101 columns]

In [123]:
X=df

In [121]:
X.isnull().sum()

0     12
1     12
2     12
3     12
4     12
      ..
95    12
96    12
97    12
98    12
99    12
Length: 100, dtype: int64

In [125]:
X.dropna(axis=0,inplace=True)

In [126]:
X.isnull().sum()

0         0
1         0
2         0
3         0
4         0
         ..
96        0
97        0
98        0
99        0
Output    0
Length: 101, dtype: int64

In [142]:
y=df['Output']
X.drop(columns=['Output'],inplace=True)

In [143]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,y, test_size=20,random_state=42)

In [ ]:
X_train.columns

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,
       72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,
       90, 91, 92, 93, 94, 95, 96, 97, 98, 99],
      dtype='object')

In [145]:
from sklearn.ensemble import RandomForestClassifier
rfc=RandomForestClassifier()

In [147]:
rfc.fit(X_train,Y_train)

RandomForestClassifier()

In [149]:
y_pred=rfc.predict(X_test)

In [150]:
from sklearn.metrics import classification_report,accuracy_score
print("Accuracy score: ",accuracy_score(Y_test,y_pred))
print(classification_report(Y_test,y_pred))

Accuracy score:  1.0
              precision    recall  f1-score   support

       False       1.00      1.00      1.00         3
        True       1.00      1.00      1.00        17

    accuracy                           1.00        20
   macro avg       1.00      1.00      1.00        20
weighted avg       1.00      1.00      1.00        20

